pkg/storage/registry/spaces/spaces.go

```diff
@@ -0,0 +1,457 @@
+// Then the gateway needs all providers below /foo/sub
+// -> only the /foo/sub provider
+//
+//           requested path   provider path
+// above   = /foo           <=> /foo/bar        -> stat(spaceid, .)    -> add metadata for /foo/bar
+// above   = /foo           <=> /foo/bar/bif    -> stat(spaceid, .)    -> add metadata for /foo/bar
+// matches = /foo/bar       <=> /foo/bar        -> list(spaceid, .)
+// below   = /foo/bar/bif   <=> /foo/bar        -> list(spaceid, ./bif)
+func (r *registry) ListProviders(ctx context.Context, filters map[string]string) ([]*registrypb.ProviderInfo, error) {
+	switch {
+	case filters["storage_id"] != "" && filters["opaque_id"] != "":
+		return r.findProvidersForResource(ctx, filters["storage_id"]+"!"+filters["opaque_id"]), nil
+	case filters["path"] != "":
+		return r.findProvidersForAbsolutePathReference(ctx, filters["path"]), nil
+	}
+	return []*registrypb.ProviderInfo{}, nil
+}
+
+// findProvidersForResource looks up storage providers based on a resource id
+// for the root of a space the res.StorageId is the same as the res.OpaqueId
+// for share spaces the res.StorageId tells the registry the spaceid and res.OpaqueId is a node in that space
+func (r *registry) findProvidersForResource(ctx context.Context, id string) []*registrypb.ProviderInfo {
+	currentUser := ctxpkg.ContextMustGetUser(ctx)
+	for address, rule := range r.c.Providers {
+		p := &registrypb.ProviderInfo{
+			Address:    address,
+			ProviderId: id,
+		}
+		filters := []*providerpb.ListStorageSpacesRequest_Filter{}
+		if rule.SpaceType != "" {
+			// add filter to id based request if it is configured
+			filters = append(filters, &providerpb.ListStorageSpacesRequest_Filter{
+				Type: providerpb.ListStorageSpacesRequest_Filter_TYPE_SPACE_TYPE,
+				Term: &providerpb.ListStorageSpacesRequest_Filter_SpaceType{
+					SpaceType: rule.SpaceType,
+				},
+			})
+		}
+		filters = append(filters, &providerpb.ListStorageSpacesRequest_Filter{
+			Type: providerpb.ListStorageSpacesRequest_Filter_TYPE_ID,
+			Term: &providerpb.ListStorageSpacesRequest_Filter_Id{
+				Id: &providerpb.StorageSpaceId{
+					OpaqueId: id,
+				},
+			},
+		})
+		spaces, err := r.findStorageSpaceOnProvider(ctx, address, filters)
+		if err != nil {
+			appctx.GetLogger(ctx).Debug().Err(err).Interface("rule", rule).Msg("findStorageSpaceOnProvider by id failed, continuing")
+			continue
+		}
+
+		if len(spaces) > 0 {
+			space := spaces[0] // there shouldn't be multiple
+			providerPath, err := rule.ProviderPath(currentUser, space)
+			if err != nil {
+				appctx.GetLogger(ctx).Error().Err(err).Interface("rule", rule).Interface("space", space).Msg("failed to execute template, continuing")
+				continue
+			}
+
+			spacePaths := map[string]string{
+				space.Id.OpaqueId: providerPath,
+			}
+			p.Opaque, err = spacePathsToOpaque(spacePaths)
+			if err != nil {
+				appctx.GetLogger(ctx).Debug().Err(err).Msg("marshaling space paths map failed, continuing")
```

Comment:
change level to `Error()` to be consistent with line 305

Reply:
Done in https://github.com/cs3org/reva/pull/2383